lib/vector-core/src/event/array.rs

```diff
@@ -304,3 +323,61 @@ impl Iterator for EventArrayIntoIter {
         }
     }
 }
+
+/// Intermediate buffer for conversion of a sequence of individual
+/// `Event`s into a sequence of `EventArray`s by coalescing contiguous
+/// events of the same type into one array. This is used by
+/// `events_into_array`.
+#[derive(Debug, Default)]
+pub struct EventArrayBuffer(Option<EventArray>);
+
+impl EventArrayBuffer {
+    #[must_use]
+    fn push(&mut self, event: Event) -> Option<EventArray> {
+        match (event, &mut self.0) {
+            (Event::Log(event), Some(EventArray::Logs(array))) => array.push(event),
+            (Event::Metric(event), Some(EventArray::Metrics(array))) => array.push(event),
+            (Event::Trace(event), Some(EventArray::Traces(array))) => array.push(event),
+            (event, current) => {
```

Comment:
It's not clear to me that this would ever be different from `(event, None)`? Or, if we add another variant would we want to replace over the top of its array?

Reply:
It could, for example, be `(Event::Log(…), Some(EventArray::Metrics(…))`

Reply:
Ah, sure, sure. Makes sense.